src/interpreter.rs

```diff
@@ -1,148 +1,128 @@
-}
-
+#[derive(Debug)]
 pub enum Instruction {
     Increment,
     Decrement,
     MoveLeft,
     MoveRight,
-    StartLoop(u32), // u32 is the index of the end of the loop
-    EndLoop(u32),   // u32 is the index of the start of the loop
+    StartLoop(usize), // usize is the index of the end of the loop
+    EndLoop(usize),   // usize is the index of the start of the loop
     Output,
     Input,
     Halt,
 }
 
+#[derive(Debug, thiserror::Error)]
+pub enum BuildLoopMapError {
+    #[error("found loops with missing close instructions at indices {0:?}")]
+    StrayOpens(Vec<usize>),
+    #[error("found loop with missing open instructions at index {0}")]
+    StrayClose(usize),
+}
+
+// Creates mapping of loops and their endpoints for easy jumping around the code at loops endpoints and startpoints
+fn build_loop_map(code: &str) -> Result<HashMap<usize, usize>, BuildLoopMapError> {
+    use BuildLoopMapError::*;
+
+    let mut loops = HashMap::new();
+    let mut open_loops = Vec::new();
+
+    for (i, c) in code.chars().enumerate() {
+        match c {
+            '[' => open_loops.push(i),
+            ']' => {
+                let start = open_loops.pop().ok_or_else(|| StrayClose(i))?;
+
+                loops.insert(start, i);
+                loops.insert(i, start);
+            }
+            _ => {}
+        }
+    }
+
+    if !open_loops.is_empty() {
+        return Err(StrayOpens(open_loops));
+    }
+
+    Ok(loops)
+}
+
+// Converts the code string into a vector of instructions
+pub fn build_instructions(code: &str, loops: &HashMap<usize, usize>) -> Vec<Instruction> {
+    let mut instructions = Vec::new();
+
+    for (i, c) in code.chars().enumerate() {
+        let instruction = match c {
+            '+' => Instruction::Increment,
+            '-' => Instruction::Decrement,
+            '<' => Instruction::MoveLeft,
+            '>' => Instruction::MoveRight,
+            '[' => Instruction::StartLoop(*loops.get(&i).unwrap()),
+            ']' => Instruction::EndLoop(*loops.get(&i).unwrap()),
+            '.' => Instruction::Output,
+            ',' => Instruction::Input,
+            _ => continue,
+        };
+
+        instructions.push(instruction);
+    }
+
+    instructions.push(Instruction::Halt);
+
+    instructions
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum NewInterpreterError {
+    #[error("error while building loop map")]
+    BuildLoopMap(#[from] BuildLoopMapError),
+}
+
 pub struct Interpreter {
     pub instructions: Vec<Instruction>,
-    pub pointer: usize,
-    pub code: String,
-    pub loops: HashMap<u32, u32>,
-    pub input: String,
-    pub tape: [u8; 3000]
+    pub instruction_pointer: usize,
+    pub data_pointer: usize,
+    pub tape: [u8; 3000],
 }
 
 impl Interpreter {
-    fn build_loop_map (&mut self) {
-        /* Creates mapping of loops and their endpoints for easy jumping around the code at loops
-         * endpoints and startpoints */
-
-        let mut open_loops: Vec<u32> = Vec::new();

```

Comment:
while verbose can be good, annotating types everywhere may add friction, in this case I suppose you picked u32 first and then wrote the rest of the code, if you had not picked a type at all, it would have defaulted to usize and the cast at line 36 would not have been required